gamestonk_terminal/portfolio/dg_api.py

```diff
@@ -1,62 +1,313 @@
+    user_token = client_details_table["data"]["id"]
+
+    # SETUP OPTIONAL CREDENTIALS
+    trading_api.credentials.int_account = int_account
+    quotecast_api = QuotecastAPI(user_token=user_token)  # noqa: F841
 
 
 def logout():
-    degiro.logout()
+    """
+    Log out from Degiro's API.
+    """
+
+    trading_api.logout()
 
 
 def show_holdings():
-    # Fetch current portfolio
-    portfolio = degiro.getdata(degiroapi.Data.Type.PORTFOLIO, True)
-
-    print("")
-    print("Stonk\t Size\t last price \t prev close \t breakeven \t % Change")
-    print("")
-
-    # Loop to get stock info based on id provided in portfolio
-    # showcase portfolio
-    for stock in portfolio:
-        if stock["positionType"] == "PRODUCT":
-            stock_info = degiro.product_info(stock["id"])
-            real_time_info = degiro.real_time_price(
-                stock["id"], degiroapi.Interval.Type.One_Day
-            )
-            size = stock["size"]
-            break_even = stock["breakEvenPrice"]
-            stonk = stock_info["symbol"]
-            last_price = real_time_info[0]["data"]["lastPrice"]
-            prev_close = real_time_info[0]["data"]["previousClosePrice"]
-            pct_change = round((last_price - prev_close) / prev_close, 3)
-            to_pr = f"{stonk}\t {size}\t {last_price}\t\t {prev_close}\t\t {break_even}\t\t {pct_change}"
-            if last_price >= prev_close:
-                print(colored(to_pr, "green"))
-            else:
-                print(colored(to_pr, "red"))
-            print("")
+    """
+    Display held products.
+    """
+
+    # FETCH HELD PRODUCTS
+    positions = fetch_current_positions()
+
+    # FORMAT DATAFRAME
+    selected_columns = [
+        "symbol",
+        "size",
+        "price",
+        "closePrice",
+        "breakEvenPrice",
+    ]
+    formatted_columns = [
+        "Stonk",
+        "Size",
+        "Last Price",
+        "Close Price",
+        "Break Even Price",
+    ]
+    fmt_positions = positions[selected_columns].copy(deep=True)
+    fmt_positions.columns = formatted_columns
+
+    fmt_positions["% Change"] = positions["price"]
+    fmt_positions["% Change"] -= fmt_positions["Close Price"]
+    fmt_positions["% Change"] /= fmt_positions["Close Price"]
+    fmt_positions["% Change"] = fmt_positions["% Change"].round(3)
+
+    # DISPLAY DATAFRAME
+    print(fmt_positions)
 
 
 def return_holdings() -> pd.DataFrame:
-    portfolio = degiro.getdata(degiroapi.Data.Type.PORTFOLIO, True)
-
-    holds = []
-    for stock in portfolio:
-        if stock["positionType"] == "PRODUCT":
-            stock_info = degiro.product_info(stock["id"])
-            yf_data = yf.Ticker(stock_info["symbol"])
-            stock["id"] = stock_info["symbol"]
-            stock["marketValue"] = (
-                yf_data.info["previousClose"] * yf_data.info["sharesOutstanding"]
-            )
-            holds.append(stock)
-    return dg_positions_to_df(holds)
+    # FETCH HELD PRODUCTS
```

Comment:
I think here you should make us of argparse.

So the function should be defined like this:

```
def return_holdings() -> pd.DataFrame:
    parser = argparse.ArgumentParser(
        add_help=False,
        prog="res",
        description="""
            ADD DESCRIPTION OF COMMAND TO USER
        """,
    )

    try:
        ns_parser = parse_known_args_and_warn(parser, other_args)
        if not ns_parser:
            return

        # CODE GOES HERE

    except Exception as e:
        print(e, "\n")
        return
```

this means that the function has both:
* try-catch block (which doesn't need to be done at the controller level)
* provides information to the user if he wants to know more about command by doing: `func -h`
*

Reply:
A sub-menu was built and the exception are caught inside the controller.
Also used `argparse` when argument where necessary.